doc/fbterm.1.in
.TH "FbTerm" "1" "@RELEASE_DATE@" "FbTerm @VERSION@" "FbTerm User Manual"
.SH "NAME"
FbTerm \- a fast Frame Buffer based TERMinal emulator for linux
.SH "SYNOPSIS"
\fBfbterm\fR [\fIoptions\fR] [\-\-] [\fIcommand\fR [\fIarguments\fR]]
.SH "DESCRIPTION"
FbTerm is a fast terminal emulator for linux with frame buffer device or VESA video card. Features include:
  
  * mostly as fast as terminal of linux kernel while accelerated scrolling is enabled
  * select font with fontconfig and draw text with freetype2, same as Qt/Gtk+ based GUI apps
  * dynamically create/destroy up to 10 windows initially running default shell
  * record scroll-back history for every window
  * auto-detect current locale and convert text encoding, support double width scripts like Chinese, Japanese etc
  * switch between configurable additional text encodings with hot keys on the fly
  * copy/past selected text between windows with mouse when gpm server is running
  * change the orientation of screen display, a.k.a. screen rotation
  * lightweight input method framework with client-server architecture
  * background image for eye candy
.SH "OPTIONS"
\fIcommand\fR [\fIarguments\fR] may be given to override the built-in choice of shell program. Normally FbTerm
checks the \fISHELL\fR variable. If that is not set, it tries to use the user's login shell program specified in the password
file. If that is not set, /bin/sh will be used. You should use the \fI\-\-\fR argument to separate FbTerm's options from
the arguments supplied to the \fIcommand\fR.

FbTerm first uses option value specified in command line arguments, then in the configure file \fI$HOME/.fbtermrc\fR.
If that file doesn't exist, FbTerm will create it with default options on startup.
.TP
\fB\-h, \-\-help\fR
display the help and exit
.TP
\fB\-V, \-\-version\fR
display FbTerm version and exit
.TP
\fB\-v, \-\-verbose\fR
display extra FbTerm's information
.TP
\fB\-n, \-\-font\-names=\fR\fItext\fR
specify font family names
.TP
\fB\-s, \-\-font\-size=\fR\fInum\fR
specify font pixel size
.TP
\fB\-W, \-\-font\-width=\fR\fInum\fR
force font width
.TP
\fB\-H, \-\-font\-height=\fR\fInum\fR
force font height
.TP
\fB\-B, \-\-font\-baseline=\fR\fInum\fR
force font baseline
.TP
\fB\-f, \-\-color\-foreground=\fR\fInum\fR
specify foreground color
.TP
\fB\-b, \-\-color\-background=\fR\fInum\fR
specify background color
.TP
\fB\-e, \-\-text\-encodings=\fR\fItext\fR
specify additional text encodings
.TP
\fB\-a, \-\-ambiguous\-wide\fR
treat ambiguous width characters as wide
.TP
\fB\-r, \-\-screen\-rotate=\fR\fInum\fR
specify the orientation of screen display
.TP
\fB\-i, \-\-input\-method=\fR\fItext\fR
specify input method program
.TP
\fB\-c, \-\-cursor\-shape=\fR\fInum\fR
specify default cursor shape
.TP
\fB\-C, \-\-cursor\-interval=\fR\fInum\fR
specify cursor flash interval
.TP
\fB\-l, \-\-term\-is\-linux\fR
set TERM variable to "linux"
.TP
\fB\-\-vesa\-mode=\fR\fInum\fR
specify VESA video mode
.TP
\fB\-\-vesa\-mode=list\fR
display available VESA video modes

.TP
see comments in \fI$HOME/.fbtermrc\fR for details of these options.
.SH "SHORTCUT SUMMARY"
keyboard:
  CTRL_ALT_E:    exit from FbTerm
  CTRL_ALT_C:    create a new window
  CTRL_ALT_D:    destroy current window
  CTRL_ALT_1:    switch to window 1
  CTRL_ALT_2:    switch to window 2
  CTRL_ALT_3:    switch to window 3
  CTRL_ALT_4:    switch to window 4
  CTRL_ALT_5:    switch to window 5
  CTRL_ALT_6:    switch to window 6
  CTRL_ALT_7:    switch to window 7
  CTRL_ALT_8:    switch to window 8
  CTRL_ALT_9:    switch to window 9
  CTRL_ALT_0:    switch to window 10
  SHIFT_LEFT:    switch to previous window
  SHIFT_RIGHT:   switch to next window
  SHIFT_PAGEUP:    history scroll up
  SHIFT_PAGEDOWN:  history scroll down
  CTRL_ALT_F1:                 switch to encoding of the current locale
  CTRL_ALT_F2 to CTRL_ALT_F6:  switch to additional encodings
  CTRL_SPACE:    toggle input method
  CTRL_ALT_K:    kill input method server

mouse:
  move when left button down:      select text
  double click with left button:   auto select text
  click with right button:         paste selected text

Sometimes above actions will not work, please try to redo them with shift key holding down.
.SH "FRAME BUFFER DEVICE"
Before executing FbTerm, make sure there is a frame buffer device in your system, and you have read/write access right
with it. Normally FbTerm tries to open /dev/fb0 and /dev/fb/0, environment variable "\fIFRAMEBUFFER\fR" may be used to override this
built-in behavior.
.SH "VESA VIDEO DEVICE"
From version 1.6, FbTerm adds VESA video card support. By default, FbTerm tries frame buffer device first, if failure, 
then tries to use VESA device with highest resolution and color depth. option "\fIvesa\-mode\fR" may be given to force only open
VESA device with the specified video mode. To see available video modes for your VESA card, execute "\fBfbterm \-\-vesa\-mode=list\fR".

Attention: 1) VESA support requires root privilege to work; 2) do not force to use VESA device on the system with frame buffer device
enabled, they maybe conflict with each other.
.SH "FONT"
FbTerm invokes fontconfig to get a font list, if the first font doesn't contain the glyph for the rendering character,
it will try second font, then the third, ... and so on, user can see this ordered font list with "\fBfbterm \-v\fR".

If you don't like the fonts selected by FbTerm, execute "fc\-list" to get available fonts, choose favorites as
the value of option "\fIfont\-names\fR". You may also modify the configure file of fontconfig, which will
change the behavior of all programs based on fontconfig!
.SH "TEXT ENCODING"
By using iconv, FbTerm converts other encodings to internal encoding UTF-8. On startup, FbTerm checks variable
\fILC_CTYPE\fR to determine the default text encoding, which is bound to shortcut CTRL_ALT_F1.
User can specify up to 5 additional encodings with option "\fItext\-encodings\fR", for example, as a Chinese user,
you set "text\-encodings=gbk,big5", shortcut CTRL_ALT_F2 will bind to gbk, and CTRL_ALT_F3 bind to big5.

In a number of CJK encodings there are ambiguous width characters which have a width of either narrow or wide
depending on the context of their use. By default, FbTerm treats them as narrow width characters,
option "\fIambiguous\-wide\fR" may be used to change the behavior.
.SH "BACKGROUND IMAGE"
FbTerm doesn't load and parse any image file with various formats directly, instead it takes a screen shot of
frame buffer device on startup if variable \fIFBTERM_BACKGROUND_IMAGE\fR is defined, then uses this screen shot as
the background for text rendering. In order to enable background image support, user should first
put an image to frame buffer device with an image viewer. A wrapper script using fbv is listed below:
\fB
.br

    #!/bin/bash

    # fbterm\-bi: a wrapper script to enable background image with fbterm
    # usage: fbterm\-bi /path/to/image fbterm-options

    echo \-ne "\\e[?25l" # hide cursor

    fbv \-ciuker "$1" << EOF
    q
    EOF

    shift
    export FBTERM_BACKGROUND_IMAGE=1
    exec fbterm "$@"
\fR    
Attention: 1) do not enable background image on frame buffer device with 8bpp depth, because FbTerm changes color map
table for correct text rendering; 2) if the screen shot is different from the original image, try to use a fast scrolling
disabled frame buffer device.
.SH "256 COLOR EXTENSION"
FbTerm supports xterm's 256 color mode extension. The first 16 colors are the default terminal colors. Additionally, there's
a 6x6x6 color cube, and 24 grayscale tones. But xterm's 256 color escape sequences conflict with the linux sequences implemented by FbTerm,
so private escape sequences were introduced to support this feature:
\fB
.br

    ESC [ 1 ; n }                   set foreground color to n (0 - 255)
    ESC [ 2 ; n }                   set background color to n (0 - 255)
    ESC [ 3 ; n ; r ; g ; b }       set color n to (r, g, b) ,  n, r, g, b all in (0 - 255)
\fR

A new terminfo database entry named "fbterm" was added to use these private sequences, all program based on terminfo should work with it.
By default, FbTerm sets environment variable "TERM" to value "fbterm". Alternatively, the "TERM" variable can be set to "linux",
in which case the 256 color mode will be unavailable.
.SH "INPUT METHOD"
Instead of adding input method directly in FbTerm, a client-server based input method framework is designed to do
this work. FbTerm acts as a client, standalone IM program as a server, and they run in separated processes.

As a normal IM user, you should install an IM program written for FbTerm, and specify it as the value of option "\fIinput\-method\fR".
CTRL_SPACE is the shortcut to activate/deactivate input method. And CTRL_ALT_K may be used to kill the IM program when it has frozen.

.SH "SECURITY NOTES"
FbTerm tries to change linux kernel key map table to setup shortcuts, which requires \fISYS_TTY_CONFIG\fR capability since kernel
version 2.6.15. It means FbTerm should be a setuid 0 program to allow non-root users to use shortcuts. FbTerm only
switches to root privilege temporarily when changing key map table, we believe it's pretty much free from security problems.
If you really don't like this and have a linux kernel with file system capabilities enabled, which allow user to give
binaries a subset of root's powers without using setuid 0 (official kernel 2.6.27 includes it), you can run command
"\fBsudo setcap 'cap_sys_tty_config+ep' /path/to/fbterm\fR".

FbTerm redirects /dev/tty0 output to the pseudo terminal of the current sub-window. In linux before version 2.6.10, anybody can do this
as long as the output was not redirected yet; since version 2.6.10, only root or a process with the \fICAP_SYS_ADMIN\fR capability may do this.
You should do similar work with above ones for FbTerm to enable this feature for non-root users. 

.SH "PERFORMANCE"
According to test results, FbTerm spends more than 95% time on painting screen. Speedup painting will significantly
increase performance. Here are some suggestions for frame buffer device:

  * Enable fast scrolling on frame buffer device. The visible screen will be just a window of the video memory,
when scrolling, FbTerm will only change the start of window, no need to repaint the whole screen.
    
  * On Intel P6 family processors the Memory Type Range Registers (MTRRs) may be used to control processor access to
memory ranges. If you have video card on a PCI or AGP bus, enabling write-combining allows bus write transfers
to be combined into a larger transfer before bursting over the PCI/AGP bus. This can increase performance of
FbTerm's paint operations.

Maybe you need to enable them manually for your video card. For example, by default fast scrolling and write-combining
are all disabled on general VESA frame buffer device, GRUB/LILO configure file should be modified to enable them.
A GRUB example is shown below:

title Ubuntu
    root (hd0,0)
    kernel /boot/vmlinuz ro root=LABEL=UBUNTU splash vga=0x317 \fBvideo=vesafb:ywrap,mtrr:3\fR
    initrd /boot/initrd.img
    
You may check fast scrolling status with "\fBfbterm \-v\fR", a message with "scrolling: redraw" means fast scrolling
is disabled, otherwise enabled.
.SH "AUTHOR"
Written by dragchan.
.SH "REPORTING BUGS"
Please submit bug reports to <\fBhttp://code.google.com/p/fbterm/issues\fR>.
.SH "COPYRIGHT"
Copyright \(co 2008-2010 dragchan.  License GPLv2: GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.  There is NO WARRANTY, to the extent permitted by law.


ChangeLog
2015-02-21 version 1.8
	* doc/fbterm.1.in:
	escape '-' (minus) with backslash in the man page
	
	* lib/shell.h lib/shell.cpp fbshell.cpp fbconfig.cpp doc/fbterm.1.in:
	allow to choose 'fbterm' or 'linux' for TERM
	
	* lib/shell.cpp doc/fbterm.1.in:
	changed the default value of TERM to 'fbterm'
	
	* fbconfig.cpp doc/fbterm.1.in:
	added short versions for most command line options
	
	* font.h font.cpp fbconfig.cpp doc/fbterm.1.in:
	added configuration option to adjust font baseline
	
	* font.cpp:
	getGlyph(): fixed incorrect glyph bitmap positioning

2010-10-05 version 1.7
	* vterm.cpp charsetmap.cpp:
	improved escape sequences support, now curses h/v lines should be drawed correctly 
	
	* fbshell.cpp:
	added redirecting /dev/tty0 output to FbTerm's sub-window
	
	* improxy.cpp input.cpp:
	added shortcut CTRL_ALT_K to kill the frozen IM server
	
	* vterm.cpp wcwidth.cpp fbconfig.cpp:
	added option "ambiguous-wide" to treat ambiguous CJK characters as wide width
	
	* font.cpp fbcofnig.cpp:
	added option "font-height" to force font height
	
	* vterm.cpp:
	do_normal_char(): fixed logical error in insert mode
	
	* shell.cpp:
	autoTextSelect(): fixed logical error
	
	* improxy.cpp:
	improved screen redrawing performance when im is active 

2009-11-14 version 1.6
	* screen.cpp screen.h screen_render.cpp:
	reorganized class Screen API for VESA support
	
	* screen_clip.cpp:
	removed this file because similar function has been recoded in improxy.cpp
	
	* vesa.cpp vesa.h vbe.h:
	added general VESA video card support
	
	* fbdev.cpp fbdev.h:
	moved frame buffer device support from class Screen to new class FbDev
	
	* immessage.h improxy.cpp:
	added IM rendering message for IM server development
	
	* input.cpp:
	moved tty initialing code from ttyInput() to switchVc()
	Ctrl+Space won't be a shortcut if user run FbTerm without input-method option
	
	* mouse.cpp:
	fixed a bug where FbTerm isn't compiled with gpm mouse support but run with gpm server enabled
	
	* fbconfig.cpp:
	added options for VESA support
	fixed some spelling errors in FbTerm's help message
	
	* doc/inputmethod.txt:
	removed content of this file to be comments of im/immessage.h and im/imapi.h
	
	* doc/fbterm.1.in:
	added description for VESA support
	fixed several spelling errors
	
	* imapi.c:
	fixed a IM dead loop bug triggered by FbTerm's crash
	fixed a bug where IM maybe doesn't process pending messages until a new message arrives

2009-04-26 version 1.5
	* fbconfig.cpp fbshell.cpp shell.cpp:
	added command line arguments to customize command executed in sub-window, which is user's login shell by default
	
	* improxy.cpp input.cpp fbshell.cpp fbshellman.cpp:
	added Alt-Fn and all FbTerm's shortcuts support when input method is actived
	
	* screen_render.cpp:
	added support for text rendering with background image
	
	* fbconfig.cpp fbshell.cpp input.cpp font.cpp screen.cpp:
	showInfo(): added option "-v/--verbose" to show extra system information
	
	* lib/shell.cpp:
	createShellProcess(): changed shell program selection algorithm
	
	* screen.cpp:
	createInstance(): changed frame buffer device selection behavior
	
	* font.cpp:
	createInstance(): removed duplicate fonts with same family from ordered font list
	
	* lib/io.cpp:
	localCodec(): fixed UTF-8 characters not showed correctly when locale is C/POSIX
	
	* lib/vterm.h:
	fixed a bug with character color for underline/halfbright attribute in version 1.4
	
	* fbterm.cpp:
	fixed a clearing screen issue on startup even in inactive tty
	
	* fbconfig.cpp:
	Config(): fixed a bug when user try to execute FbTerm in initscripts and variable HOME is not defined
	
	* input.cpp:
	fixed a bug that ctrl/space release keycodes are sent to im server when user enable it by pressing ctrl+space

2009-03-14 version 1.4
	* screen_render.cpp:
	improved text rendering performance
	
	* lib/vterm_state.cpp lib/vterm_action.cpp fbshell.cpp:
	added private escape sequences for 256 color support 
	
	* font.cpp fbconfig.cpp:
	added a option to adjust font width
	
	* fbterm.cpp fbio.cpp:
	added using select() for old 2.2/2.4 kernel
	
	* font.cpp:
	fixed a crash bug where font's width/height returns zero for bitmap fonts
	
	* font.cpp screen.cpp:
	removed codes for non fixed width font
	
	* configure:
	fixed a failure with cross-compiling environment

2008-12-20 version 1.3
	* fbconfig.cpp:
	added command line arguments to change option values

	* fbterm.cpp fbshellman.cpp fbshell.cpp improxy.cpp improxy.h input.cpp im/*:
	added client-server based input method framework
		
	* screen.cpp screen_clip.cpp:
	added screen clip support for IM
	
	* screen.cpp font.cpp:
	added screen rotation support
	
	* screen.cpp:
	drawGlyph(): fixed a spelling error of '=' as '-', which even appears in version 1.0
	move(): fixed a duplicate repaint bug while fast scrolling type is ypan
	moved paint code to a separate source file screen_render.cpp
	
	* screen_render.cpp:
	setupSysPalette(): added support visual FB_VISUAL_DIRECTCOLOR
	
	* mouse.cpp lib/shell.cpp lib/shell.h:
	added mouse wheel support
	
	* lib/vterm.h lib/vterm.cpp:
	decreased memory usage of every vterm instance by changing struct CharAttr size from 4 to 2 bytes
	
	* font.cpp:
	replaced hash table based glyph cache with direct array index, removed dependence of hash.h/hash.cpp
	
	* input.cpp:
	readyRead(): fixed a unicode input character processing error
	
	* lib/shell.cpp lib/vterm.cpp:
	fixed a bug of not inverse color for selected text when switch back to the shell
	moved mouse pointer drawing code to class FbShell
	
	* fbshell.cpp:
	fixed a bug of maybe not restore original console state after FbTerm exited
	moved class FbShellManager to separate source files fbshellman.h and fbshellman.cpp
	
	* fbshellman.cpp:
	switchShell(): fixed a duplicate repaint bug while switch to same window
	
	* configure src/Makefile.in fbterm.cpp fbshell.cpp input.cpp:
	added using filesystem capability attributes offered by kernel 2.6.27, instead of setting set-user-ID bit on FbTerm


src/fbterm.cpp
/*
 *   Copyright © 2008-2010 dragchan <zgchan317@gmail.com>
 *   This file is part of FbTerm.
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <unistd.h>
#include <stdio.h>
#include <signal.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <linux/vt.h>
#include <linux/kdev_t.h>
#include "config.h"
#include "fbterm.h"
#include "fbshell.h"
#include "fbshellman.h"
#include "fbconfig.h"
#include "fbio.h"
#include "screen.h"
#include "input.h"
#include "input_key.h"
#include "mouse.h"

#ifndef WAIT_ANY
#define WAIT_ANY (-1)
#endif

#ifdef HAVE_SIGNALFD
#include <sys/signalfd.h>

static sigset_t oldSigmask;

class SignalIo : public IoPipe {
public:
	SignalIo(sigset_t &sigmask);

private:
	virtual void readyRead(s8 *buf, u32 len);
};

SignalIo::SignalIo(sigset_t &sigmask)
{
	int fd = signalfd(-1, &sigmask, 0);
	setFd(fd);
}

void SignalIo::readyRead(s8 *buf, u32 len)
{
	signalfd_siginfo *si = (signalfd_siginfo*)buf;
	for (len /= sizeof(*si); len--; si++) {
		FbTerm::instance()->processSignal(si->ssi_signo);
	}
}
#else

static volatile sig_atomic_t pendsigs = 0;

static void signalHandler(s32 signo)
{
	pendsigs |= 1 << signo;
}

static void pollSignal()
{
	if (!pendsigs) return;

	sig_atomic_t sigs = pendsigs;
	pendsigs = 0;

	for (u32 i = 0; i < sizeof(sigs); i++) {
		u8 sig8 = (sigs >> (i * 8)) & 0xff;
		if (!sig8) continue;

		for (u32 j = 0; j < 8; j++) {
			if (sig8 & (1 << j)) {
				FbTerm::instance()->processSignal(i * 8 + j);
			}
		}
	}
}
#endif

DEFINE_INSTANCE_DEFAULT(FbTerm)

FbTerm::FbTerm()
{
	mInit = false;
	init();
}

FbTerm::~FbTerm()
{
	IoDispatcher::uninstance();
	FbShellManager::uninstance();
	Screen::uninstance();
}

void FbTerm::init()
{
	if (!TtyInput::instance() || !Screen::instance()) return;

	struct vt_mode vtm;
	vtm.mode = VT_PROCESS;
	vtm.waitv = 0;
	vtm.relsig = SIGUSR1;
	vtm.acqsig = SIGUSR2;
	vtm.frsig = 0;
	ioctl(STDIN_FILENO, VT_SETMODE, &vtm);

#ifdef HAVE_SIGNALFD
	sigset_t sigmask;
	sigemptyset(&sigmask);

	sigaddset(&sigmask, SIGCHLD);
	sigaddset(&sigmask, SIGUSR1);
	sigaddset(&sigmask, SIGUSR2);
	sigaddset(&sigmask, SIGALRM);
	sigaddset(&sigmask, SIGTERM);
	sigaddset(&sigmask, SIGHUP);

	sigprocmask(SIG_BLOCK, &sigmask, &oldSigmask);
	new SignalIo(sigmask);
#else
	sighandler_t sh = signalHandler;

	signal(SIGCHLD, sh);
	signal(SIGUSR1, sh);
	signal(SIGUSR2, sh);
	signal(SIGALRM, sh);
	signal(SIGTERM, sh);
	signal(SIGHUP, sh);
#endif
	signal(SIGPIPE, SIG_IGN);

	Mouse::instance();
	mInit = true;
}

static bool isActiveTerm()
{
	struct vt_stat vtstat;
	ioctl(STDIN_FILENO, VT_GETSTATE, &vtstat);

	struct stat ttystat;
	fstat(STDIN_FILENO, &ttystat);

	return vtstat.v_active == MINOR(ttystat.st_rdev);
}

void FbTerm::run()
{
	if (!mInit) return;

	if (isActiveTerm()) processSignal(SIGUSR2);

	FbShellManager::instance()->createShell();

	mRun = true;
	FbIoDispatcher *io = (FbIoDispatcher*)IoDispatcher::instance();
	while (mRun) {
		io->poll();
#ifndef HAVE_SIGNALFD
		pollSignal();
#endif
	}

	if (isActiveTerm()) processSignal(SIGUSR1);
}

void FbTerm::processSignal(u32 signo)
{
	switch (signo) {
	case SIGTERM:
	case SIGHUP:
		exit();
		break;

	case SIGALRM:
		FbShellManager::instance()->drawCursor();
		break;

	case SIGUSR1:
		FbShellManager::instance()->switchVc(false);
		Screen::instance()->switchVc(false);
		TtyInput::instance()->switchVc(false);
		Mouse::instance()->switchVc(false);
		ioctl(STDIN_FILENO, VT_RELDISP, 1);
		break;

	case SIGUSR2:
		Mouse::instance()->switchVc(true);
		TtyInput::instance()->switchVc(true);
		Screen::instance()->switchVc(true);
		FbShellManager::instance()->switchVc(true);
		break;

	case SIGCHLD:
		if (mRun) {
			s32 pid = waitpid(WAIT_ANY, 0, WNOHANG);
			if (pid > 0) {
				FbShellManager::instance()->childProcessExited(pid);
			}
		}
		break;

	default:
		break;
	}
}

void FbTerm::processSysKey(u32 key)
{
	FbShellManager *manager = FbShellManager::instance();

	switch (key) {
	case CTRL_ALT_E:
		exit();
		break;

	case SHIFT_PAGEDOWN:
	case SHIFT_PAGEUP:
		manager->historyScroll(key == SHIFT_PAGEDOWN);
		break;

	case CTRL_ALT_C:
		manager->createShell();
		break;

	case CTRL_ALT_D:
		manager->deleteShell();
		break;

	case CTRL_ALT_1 ... CTRL_ALT_0:
		manager->switchShell(key - CTRL_ALT_1);
		break;

	case SHIFT_LEFT:
		manager->prevShell();
		break;

	case SHIFT_RIGHT:
		manager->nextShell();
		break;

	case CTRL_ALT_F1 ... CTRL_ALT_F6:
		if (manager->activeShell()) {
			manager->activeShell()->switchCodec(key - CTRL_ALT_F1);
		}
		break;

	case CTRL_SPACE:
		if (manager->activeShell()) {
			manager->activeShell()->toggleIm();
		}
		break;

	case CTRL_ALT_K:
		if (manager->activeShell()) {
			manager->activeShell()->killIm();
		}

	default:
		break;
	}
}

void FbTerm::initChildProcess()
{
	setuid(getuid());

#ifdef HAVE_SIGNALFD
	sigprocmask(SIG_SETMASK, &oldSigmask, 0);
#endif

	signal(SIGPIPE, SIG_DFL);
}

int main(int argc, char **argv)
{
	seteuid(getuid());

	if (Config::instance()->parseArgs(argc, argv)) {
		FbTerm::instance()->run();
		FbTerm::uninstance();
	}

	Config::uninstance();
	return 0;
}
